libraries/tock-register-interface/src/peripheral.rs

```diff
@@ -0,0 +1,278 @@
+/// The doc comments will be emitted on the trait declaration, so they show up
+/// in rustdoc.
+///
+/// # Blanket Deref impl
+/// `tock_registers::peripheral!` emits a blanket implementation of the
+/// peripheral trait for types that `Deref` to an implementation of the
+/// peripheral:
+/// ```ignore
+/// impl<T: core::ops::Deref> Foo for T where T::Target: Foo {
+///     ...
+/// }
+/// ```
+/// Using this blanket impl, a driver that looks like:
+/// ```ignore
+/// struct FooDriver<F: Foo> {
+///     foo: F,
+/// }
+/// ```
+/// Can be used as either a `FooDriver<&FakeFoo>` or a `FooDriver<Rc<FakeFoo>>`
+/// in unit tests.
+///
+/// # Bus Adapters
+/// If you add the `#[allow_bus_adapter]` attribute to a peripheral:
+/// ```
+/// tock_registers::peripheral! {
+///     #[allow_bus_adapter]
+///     Foo {
+///         0x0 => ctrl: u16 { Write },
+///         // Since the offset of register changes depending on the passed bus
+///         // adapter, you can have the offset be inferred by specifying _ as
+///         // the offset:
+///         _ => status: u8 { Read },
+///     }
+/// }
+/// ```
+/// the generated `RealFoo` structure will gain a
+/// `B: tock_registers::BusAdapter` generic argument:
+/// ```ignore
+/// #[derive(Clone, Copy)]
+/// struct RealFoo<
+///     P: tock_registers::MmioPointer,
+///     B: tock_registers::BusAdapter<u8> = tock_registers::DirectBus>
+///     { ... }
+/// ```
+/// This argument is used to support unusual chip designs (mainly some LiteX
+/// designs).
+///
+/// The above example also showed inferred offsets, which are valid in other
+/// contexts as well. Note that a padding field cannot be followed by a field
+/// with an inferred offset.
+///
+/// # Generated `macro_rules!` macro
+/// In addition to the `Foo` trait and `RealFoo` trait, `peripheral!` also
+/// generates a `macro_rules!` macro with the same name as the trait:
+/// ```ignore
+/// #[macro_export]
+/// macro_rules! Foo { ... }
+/// ```
+/// This is currently unused, but is emitted because we may want to use it in
+/// the future, and adding it later would be a breaking change.
+///
+/// # Unsafe operations
+/// Registers can be unsafe to use:
+/// ```
+/// tock_registers::peripheral! {
+///     Foo {
+///         // A write-only register that is unsafe to write.
+///         0x0 => a: u8 { UnsafeWrite },
+///
+///         // A register that can safely be read from or unsafely be written to
+///         0x1 => b: u8 { Read, UnsafeWrite },
+///
+///         // A register can also be unsafe to read (for use if reading it
+///         // triggers a memory-unsafe operation).
+///         0x2 => c: u8 { UnsafeRead },
+///
+///         // Any combination of readable, writable, safe, and unsafe is
+///         // possible.
+///     }
+/// }
+/// ```
+/// These registers implement the [`UnsafeRead`] and [`UnsafeWrite`] traits
+/// instead of their safe equivalents.
+///
+/// # `cfg` attributes
+/// Peripheral declarations can have `cfg` attributes:
```

Comment:
We have a set of attribute-position proc macros that yield `cfg` attributes - e.g. `#[cfg_device]` and `#[cfg_host]`. Would they work in this position?

Reply:
It's fine if they can't work here, as we have simple workarounds - but the lack of support for custom attributes should be documented if so